src/eviction/lru.c

```diff
@@ -224,49 +224,88 @@ static void balance_lru_list(ocf_cache_t cache,
 
 /*-- End of LRU functions*/
 
-void evp_lru_init_cline(ocf_cache_t cache, ocf_cache_line_t cline)
-{
-	struct lru_eviction_policy_meta *node;
-
-	node = &ocf_metadata_get_eviction_policy(cache, cline)->lru;
-
-	node->hot = false;
-	node->prev = end_marker;
-	node->next = end_marker;
-}
-
-static struct ocf_lru_list *evp_lru_get_list(struct ocf_user_part *part,
+static struct ocf_lru_list *evp_lru_get_list(struct ocf_part_runtime *part,
 		uint32_t evp, bool clean)
 {
-	return clean ? &part->runtime->eviction[evp].policy.lru.clean :
-			&part->runtime->eviction[evp].policy.lru.dirty;
+	return clean ? &part->eviction[evp].policy.lru.clean :
+			&part->eviction[evp].policy.lru.dirty;
 }
 
 static inline struct ocf_lru_list *evp_get_cline_list(ocf_cache_t cache,
 		ocf_cache_line_t cline)
 {
-	ocf_part_id_t part_id = ocf_metadata_get_partition_id(cache, cline);
-	struct ocf_user_part *part = &cache->user_parts[part_id];
 	uint32_t ev_list = (cline % OCF_NUM_EVICTION_LISTS);
+	ocf_part_id_t part_id;
+	struct ocf_part_runtime *part;
+
+	part_id = ocf_metadata_get_partition_id(cache, cline);
+	ENV_BUG_ON(part_id == PARTITION_UNSPECIFIED ||
+			part_id == PARTITION_INVALID);
+	part = cache->user_parts[part_id].runtime;
 
 	return evp_lru_get_list(part, ev_list,
 			!metadata_test_dirty(cache, cline));
 }
 
+static void evp_lru_move(ocf_cache_t cache, ocf_cache_line_t cline,
+		struct ocf_part_runtime *src_part, struct ocf_lru_list *src_list,
+		struct ocf_part_runtime *dst_part, struct ocf_lru_list *dst_list,
+		ocf_part_id_t dst_part_id)
+{
+	remove_lru_list(cache, src_list, cline);
+	balance_lru_list(cache, src_list);
+	add_lru_head(cache, dst_list, cline);
+	balance_lru_list(cache, dst_list);
+	env_atomic_dec(&src_part->curr_size);
+	env_atomic_inc(&dst_part->curr_size);
+	ocf_metadata_set_partition_id(cache, cline, dst_part_id);
```

Comment:
That's a good point. Even though balancing is pretty cheap, the freelist is used really often (and this is for both - eviction and mapping), so it may be worth to skip balancing for freelist. This btw can be done easily by adding `is_balanced` flag to part and setting it to `false` on freelist, and then just early returning from `balance_lru_list()` if the flag is set.

Reply:
done